cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/AwsSubnetIgwExplorer.java

```diff
@@ -68,4 +86,140 @@ public boolean hasInternetGatewayOfSubnet(List<RouteTable> routeTables, String s
+        List<Tag> tags = awsSubnets.stream()
+            .filter(subnet -> subnetId.equalsIgnoreCase(subnet.getSubnetId()))
+            .flatMap(subnet -> subnet.getTags().stream())
+            .collect(Collectors.toList());
+        LOGGER.debug("Found tags {} for subnet {}", tags, subnetId);
+        return tags.stream()
+            .anyMatch(tag ->
+                ENDPOINT_GATEWAY_OVERRIDE_KEY.equalsIgnoreCase(tag.getKey()) &&
+                    ENDPOINT_GATEWAY_OVERRIDE_VALUE.equalsIgnoreCase(tag.getValue())
+            );
+    }
+
+    private boolean isRoutableToFirewallWithInternetGateway(List<RouteTable> routeTables, String subnetId, String vpcId,
+            AmazonNetworkFirewallClient nfwClient) {
+        boolean result = false;
+        try {
+            List<String> gatewayIds = getPotentialFirewallEndpointIds(routeTables, subnetId, vpcId);
+            if (!gatewayIds.isEmpty()) {
+                LOGGER.info("Found gateway ids {} in route table for subnet {}", gatewayIds, subnetId);
+                List<FirewallMetadata> firewallMetadata = getAllFirewallMetadataForVpc(nfwClient, vpcId);
+                if (!firewallMetadata.isEmpty()) {
+                    LOGGER.info("Found firewalls {} in VPC {}", firewallMetadata.stream().map(FirewallMetadata::getFirewallName), vpcId);
+                    result = hasRouteToPublicFirewall(firewallMetadata, routeTables, subnetId, vpcId, nfwClient, gatewayIds);
+                } else {
+                    LOGGER.info("Did not find any firewalls in VPC {}", vpcId);
+                }
+            } else {
+                LOGGER.info("Did not find any potential firewall routing paths for subnet {}", subnetId);
+            }
+        } catch (AWSNetworkFirewallException e) {
+            if (ACCESS_DENIED.equals(e.getErrorCode())) {
+                LOGGER.info("User does not have permission to query firewall resources: {}. Skipping firewall IGW check.",
+                    e.getMessage());
+            } else {
+                LOGGER.warn("Unexpected exception while fetching firewall information. Skipping firewall IGW check.", e);
+            }
+        }
+        return result;
+    }
+
+    private List<String> getPotentialFirewallEndpointIds(List<RouteTable> routeTables, String subnetId, String vpcId) {
+        List<String> gatewayIds = new ArrayList<>();
+        LOGGER.info("Searching subnet {} route table for potential firewall routing", subnetId);
+        Optional<RouteTable> subnetRouteTable = getRouteTableForSubnet(routeTables, subnetId, vpcId);
+        LOGGER.info("Searching subnet route table for any routes that start with the firewall gateway prefix ({})", FW_PREFIX);
+        subnetRouteTable.ifPresent(
+            routeTable ->
+                gatewayIds.addAll(subnetRouteTable.get().getRoutes().stream()
+                    .filter(route -> StringUtils.isNotEmpty(route.getGatewayId()) && route.getGatewayId().startsWith(FW_PREFIX))
```

Comment:
I am not sure if other cloudbreak code operations this way or not. But I have a minor concern that using the prefix might catch things we don't intend. Maybe this is a common practice in cloudbreak, if so then I am OK with it.